### ISSUE TITLE
false positive with no-unnecessary-type-assertion

### ISSUE THREAD

**Issue by Mister-Hope** [oxc-project/tsgolint#1047]
Example code here:

```ts
// oxlint-disable-next-line typescript/no-explicit-any
type IAnyObject = Record<any, any>;

const { isArray } = Array;

// oxlint-disable-next-line typescript/no-unnecessary-type-parameters
const isPlainObject = <InferredType extends IAnyObject = IAnyObject>(
  val: unknown,
): val is InferredType => Object.prototype.toString.call(val) === "[object Object]";

/**
 * Deep merge objects to the first one
 *
 * @param originObject Object to be merged into
 * @param overrideObjects Objects to merge into the first one
 * @returns Merged object
 */
export const deepAssign = <
  TargetObject extends IAnyObject,
  MergedObject extends IAnyObject = TargetObject,
  // oxlint-disable-next-line typescript/no-unnecessary-type-parameters
  FinalObject extends Partial<TargetObject> & Partial<MergedObject> = TargetObject & MergedObject,
>(
  originObject: TargetObject,
  ...overrideObjects: (MergedObject | null | undefined)[]
): FinalObject => {
  if (overrideObjects.length === 0) return originObject as unknown as FinalObject;

  /** Object being merged */
  const assignObject = overrideObjects.shift();

  if (assignObject) {
    Object.entries(assignObject).forEach(([property, value]) => {
      if (property === "__proto__" || property === "constructor") return;
      if (isPlainObject(originObject[property]) && isPlainObject(value)) {
        deepAssign(originObject[property], value);
      } else if (isArray(value)) {
        (originObject as IAnyObject)[property] = [...(value as unknown[])];
      } else if (isPlainObject(value)) {
        (originObject as IAnyObject)[property] = {
          ...value,
        };
      } else {
        (originObject as IAnyObject)[property] = assignObject[property] as unknown;
      }
    });
  }

  return deepAssign(originObject, ...overrideObjects);
};
```

[playground link](https://www.typescriptlang.org/play/?#code/PTAEHsA8BsEsDsAuBaAJrAzgQwEbQKbLz6QpzGiICeADvhgMYBOsNiw84yJNcDsKLPCoAoanVABJAILCA8jgBW+BolABeUACUV4JqgA8QqgBpQxgHwBuESIbh4GNQG9QmaUyZYqoAL4bQDy8qGxEQCBgEFHRsPEJiUmRyfEpaemZWdk5kAFd4YgZ6bCYqZHFCGiwvAFt8RHwmDDsHJzcMAAVoLAQFZVUAg0l4ADMGpnxUABU00BJ6+FQMKVkqXpU1TRl5JXWLAAoRUFAANyxoAC5QPIBrTgB3eBMRAEpL0+g2qRGxiemJdQsoDWqgAdDQmOBEJC0iCoQBlRAseAAcxBDDO0D272eGnUmgARABtcA7frAxAAXXxoWAACpaYdaaAACL4fA0UC1JjIlIkvqIJZQygACxSw1gjTUDnwjMZoAAApUahAWMieqS1OTKOBQDgUlyeag3EhwHLFVUsNUIMcxrBUPhyUtHdrOQ0ecahYhRaBxZKIMQzeNEDkmI5QABZN0TCAaxnAEQ8PRqeyONT29nSDAYWDI+ADQ6gSZVHmILVzfALJZbVYap5HSPciZl0gVxbLbb8gJFxul2sF8JQcjRTC4AhEOZJBApcqMFhsDhcPIFIpVUrlZBKy11BpNI4AMQQZ2b8zb7SqiFgZwM3ZL5MBADJQGemBerw3DXeu8W6lrH++m32+wFnoObqvylw3j+fZHCCsHgDanh2g6GoYJcez-qgWoAD6gPAOTQB8OF5Pa4rEKgzyEhSLyXAe8BHhqGiAs4BawMMoB7PBtr2o6IIECiXq4poAAMOJBiGeYgWq8BalgSw3PceayaAtH0fyoRHHSTJanqCDIq6jZGrS8ZHCmrSydmuZaponGIdxKEghgwqsYgezPOpbhsXs5k5tJGo4sxRxHOSIIVoisD0F5WY+eSzwgsMegAKJYAwwp7HshLguAdAvqYJxnDk+AUjiAKgAFgVHKx7GZdl1CCaA+IAPoNZlUJNfioBYTh1UNLVeIEqZiI5Koej4qJdTiTY5UVZ5mCdN0vn8hxqpgesGUQjVVBFaA96PrNXQraoWL5fgzz+QWU2gOmNCZhZ8BLaBC2rd1OUUmY7wFW552Bf4+DQBgKSVXs7ieN4R3QB9Z0XUc91STJVYrDFa1ZT1m0BISsEgmDBXmHJ8C3OADyUc8FKTVNP1-QDM0dPtj2He9J2Q1DMMHWoSnVojz3UBSARlVDoAY-TdZQ74pPleT-2lV95XM7TrPwx26wUZzqOaN5lkakjG3c0p8kE-AovfV9vifUcvi2EcYmhpdbLXVFuYy+SZgYzZLB2fyGCfSLIhAA) shows no errors

oxlint-tsgolint now gives:

```
  ⚠ typescript(no-unnecessary-type-assertion): This assertion is unnecessary since the receiver accepts the original type of the expression.
    ╭─[packages/create/src/utils/deepAssign.ts:27:43]
 26 │ ): FinalObject => {
 27 │   if (overrideObjects.length === 0) return originObject as unknown as FinalObject;
    ·                                           ───────────────────────────────────────
 28 │ 
    ╰────

  ⚠ typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
    ╭─[packages/create/src/utils/deepAssign.ts:38:23]
 37 │       } else if (isArray(value)) {
 38 │         (originObject as IAnyObject)[property] = [...(value as unknown[])];
    ·          ──────┬───── ──────┬───────
    ·                │            ╰── Casting it to 'IAnyObject' is unnecessary
    ·                ╰── This expression already has the type 'TargetObject'
 39 │       } else if (isPlainObject(value)) {
    ╰────

  ⚠ typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
    ╭─[packages/create/src/utils/deepAssign.ts:40:23]
 39 │       } else if (isPlainObject(value)) {
 40 │         (originObject as IAnyObject)[property] = {
    ·          ──────┬───── ──────┬───────
    ·                │            ╰── Casting it to 'IAnyObject' is unnecessary
    ·                ╰── This expression already has the type 'TargetObject'
 41 │           ...value,
    ╰────

  ⚠ typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
    ╭─[packages/create/src/utils/deepAssign.ts:44:23]
 43 │       } else {
 44 │         (originObject as IAnyObject)[property] = assignObject[property] as unknown;
    ·          ──────┬───── ──────┬───────
    ·                │            ╰── Casting it to 'IAnyObject' is unnecessary
    ·                ╰── This expression already has the type 'TargetObject'
 45 │       }
    ╰────
```

However, after auto fixing:

[ts playground](https://www.typescriptlang.org/play/?#code/PTAEHsA8BsEsDsAuBaAJrAzgQwEbQKbLz6QpzGiICeADvhgMYBOsNiw84yJNcDsKLPCoAoanVABJAILCA8jgBW+BolABeUACUV4JqgA8QqgBpQxgHwBuESIbh4GNQG9QmaUyZYqoAL4bQDy8qGxEQCBgEFHRsPEJiUmRyfEpaemZWdk5kAFd4YgZ6bCYqZHFCGiwvAFt8RHwmDDsHJzcMAAVoLAQFZVUAg0l4ADMGpnxUABU00BJ6+FQMKVkqXpU1TRl5JXWLAAoRUFAANyxoAC5QPIBrTgB3eBMRAEpL0+g2qRGxiemJdQsoDWqgAdDQmOBEJC0iCoQBlRAseAAcxBDDO0D272eGnUmgARABtcA7frAxAAXXxoWAACpaYdaaAACL4fA0UC1JjIlIkvqIJZQygACxSw1gjTUDnwjMZoAAApUahAWMieqS1OTKOBQDgUlyeag3EhwHLFVUsNUIMcxrBUPhyUtHdrOQ0ecahYhRaBxZKIMQzeNEDkmI5QABZN0TCAaxnAEQ8PRqeyONT29nSDAYWDI+ADQ6gSZVHmILVzfALJZbVYap5HSPciZl0gVxbLbb8gJFxul2sF8JQcjRTC4AhEOZJBApcqMFhsDhcPIFIpVUrlZBKy11BpNI4AMQQZ2b8zb7SqiFgZwM3ZL5MBADJQGemBerw3DXeu8W6lrH++m32+wFnoObqvylw3j+fZHCCsHgDanh2g6GoYJcez-qgWoAD6gPAOTQB8OF5Pa4rEKgzyEhSLyXAe8BHhqGiAs4BawMMoB7PBtr2o6IIECiXq4poAAMOJBiGeYgWq8BalgSy0fR-KhEcdJMlqeoIMirqNkatLxkcKatLJ2a5lqmicYh3EoSCGDCqxiB7M8SluGxexGTm0kajizFHEc5IghWiKwPQrlZu55LPCCwx6AAolgDDCnseyEuC4B0C+pgnGcOT4BSOIAqA3k+UcrHsSlaXUIJoD4gA+tVKVQrV+KgFhOFlQ0FV4gSBmIjkqh6Piol1OJNhFcVLmYJ03QefyHGqmB6zJRC5VULloD3o+E1dPNqhYll+DPF5BajaA6Y0JmxnwLNoHTQtbXpRSZjvNljlHT5-j4NAGApCVezuJ43i7dAz2HcdRySdtiCLal7UrQEhKwSCgPZeYSw3Pc8CUc8FIjaN72fd940dFtN07U9+0g6D4Mk5Dd3UBSASFaDoAI2Tdag74ONFXjX0Fa9RVU+SUPLfTmhuSZGpCzD9OyVc8C3OADyc29r2+C9Ry+LYRxiaGJ1smdoW5ldUnkmYCPmSwln8hgL0cyIQA)

The 4 changes leads to 4 errors:

```
Conversion of type 'TargetObject' to type 'FinalObject' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  'TargetObject' is assignable to the constraint of type 'FinalObject', but 'FinalObject' could be instantiated with a different subtype of constraint 'Partial<TargetObject> & Partial<MergedObject>'.
Type 'TargetObject' is generic and can only be indexed for reading.
Type 'TargetObject' is generic and can only be indexed for reading.
Type 'TargetObject' is generic and can only be indexed for reading.
```

**Comment by Mister-Hope** [comment]
This happens in many parts, which the `as XXX` cast definitely narrows / loose the type to achieve some works, but oxlint-tsgolint thinks not.

It seems all of them has something related to generic types.

**Comment by Mister-Hope** [comment]
Probably the same as https://github.com/oxc-project/tsgolint/issues/1044

**Comment by Mister-Hope** [comment]
Hi @camc314 , I don't think this is a feature, 0.23 works fine totally.

This rule is important, so it's not affordable for me to disable the whole rule.

**Comment by camc314** [comment]
sorry yeah I just tagged this wrong, wasn't deliberate
